### ISSUE TITLE
Message 90005 makes [AV]menu return to the root menu instead of current submenu

### ISSUE THREAD

**Issue by Sei-Lisa** [AVsitter/AVsitter#77]
Message 90005 makes [AV]menu return to the root menu instead of the current submenu

To reproduce:

- Create an AVpos notecard with these contents:
```
    TOMENU submenu
    MENU submenu
    BUTTON re-show|90005
```
- Drop it and [AV]menu into a prim.
- Click on the object, select "submenu" then "re-show". Observe that it returns to the root menu instead of the currently active menu.

This returns to the currently active menu in all versions before 2.2.

Per report by Synsational Difference.

Editing [AV]menu as follows fixes the second issue for me: https://github.com/Sei-Lisa/AVsitter/commit/avmenu-stay-in-current-menu ; however, I'd like @codeviolet to weigh in as to whether that's all that is needed, or if this fix alone causes other problems and further action is needed.

**Comment by codeviolet** [comment]
It's a deceptively tricky one...

In response to a user request just prior to the open-sourse release (which was for 90005 to be subject to [AV]menu's security check -  which they wanted for some reason), the way 90005 is handled by [AV]menu was changed to be sent through the ```menu_check()``` function, the same as if the object had been 'touched'. Before that 90005 would just open the menu directly, bypassing the security check in ```menu_check()``` and bypassing the reset of ```current_menu``` to ```-1```, therefore just opening whatever menu was selected last. Changing this behavior seems to have introduced the bug.

I don't think it's enough to only remove the ```current_menu = -1;``` from ```menu_check()``` as this would also mean that 'touching' the object would bring up whatever submenu was open last, even if touched by a new user, or at a later time.

My thoughts for how [AV]menu *should* behave here:

* 'touching' the object brings up the TOP menu, regardless of what submenu was being used previously.
* 90005 brings up the LAST menu used (unless a specific submenu is requested by the string sent with 90005, per feature request at #75).
* In addition, scripters (like the one the buggy change was made for) will now need a new way to bring up the TOP menu. So I suggest [AV]menu use [90004](https://avsitter.github.io/avsitter2_scripting.html#message-90004) for this, which already exists for [AV]sitA/B. If the fix is implemented as I've described I can contact the person the change was made for, and inform them to use 90004 in their script instead of 90005, if they need to bring up the TOP menu.

If that makes sense, I'll try to put together a PR for this change (including #75). I think the only change to the instructions would be the additional mention that 90004 is also available to [AV]menu.

**Comment by Sei-Lisa** [comment]
I'm examining AVsitB and there doesn't seem to be a similar check, so that is a slight behaviour change between both. I see uses for bypassing security, e.g. using an external access control method instead of the built-in one.

For that reason, and to keep compatibility between [AV]sitB and [AV]menu, I think that the access-checked menus should have different codes, say 90104 and 90105. Those codes could also be implemented in [AV]root-security in future.

**Comment by codeviolet** [comment]
Keeping in mind the specific user issue that got us here, I've just submitted a PR that addresses this bug and also allows a submenu to be specified in the link message string. I believe these changes will address this bug and the original issue #75.

I'm not convinced that the API should provide a method to bypass AVmenu's inbuilt security: I worry that providing a method to bypass the security is likely to lead to user difficulty on occasion as it had for the person for whom the initial buggy change was made (i.e: "why is it ignoring my security setting?").

With sitA/B, I don't think it was ever intended to provide ways of bypassing the root-security settings (although maybe there is good reason to?). It's largely the way it is because of the need for sitA/B to work without the root-security script. 90005 is used both as an "internal" link message as well as public API - perhaps this is a problem to unravel eventually.

If you see a need for them, the new link messages 90104 & 90105 that you suggest could be added in separately in the future, along with equivalents to sitA/B. Yet, with a total of 4 different link message numbers that do essentially the same thing, I worry that's feature bloat... In the spirit of keeping the instructions succinct I would hope that 90004 & 90005 would be somehow combined into a single link message before multiplying the options (if offering both access-checked and non-checked variants in future).

Hopefully I have not messed up things by submitting the PR :) My hope was to prioritize fixes for the issues the user was having.

**Comment by Sei-Lisa** [comment]
tl;dr: Make 90005 behave as before, and add a notification when security settings change.

Long version:

If we make 90005 behave differently in [AV]menu and [AV]sitA/B, that needs to be documented as a difference. Furthermore, this behaviour change is not necessarily an improvement, because it hinders the implementation of external security, which needs bypassing of built-in security, similarly to how [AV]root-security does it. I see this as a regression, because of its potential to break existing setups or expectations.

We can document 90005 as behaving differently in [AV]menu with respect to [AV]sitA/B, but if/when we're requested a way to bypass the built-in security in order to implement custom security, what will we do? We will have two options:

  1. Create a new 90105 message that bypasses built-in security in [AV]menu.
  2. Revert 90005 to bypass security and add 90105 to check built-in security.

(2) is of course undesirable, because it's yet another behaviour change when one has already been established. But it may be necessary. Imagine this scenario. I also see potential for uses wanting [AV]sitA/B to be checked for built-in security when using [AV]root-security and triggering the menu through a link message. What will we do then? If we follow (1) above, we would need to create a 90105 message in [AV]root-security that checks for security, and document that they behave in opposite ways between [AV]menu and [AV]sitA/B, which is a mess.  We can also have a message code other than 90105 to prevent a complete reversal, but that doesn't really improve the situation. If we follow (2), all will be fine, because [AV]menu will be consistent with [AV]sitA/B.

But why not just revert 90005 *now* that the new behaviour is still fairly new, to do what it did pre-2.2, and implement this user's request as a new message, and get things right from now on? The only documentation change needed will be the addition of 90105.

I understand your concern about feature bloat. But, well, it's what has made AVsitter popular in the first place, and most users don't use even half its features. This particular feature is also not a big concern memory wise. [AV]menu has way more memory free than [AV]sitB, and I don't expect this addition to change that situation significantly. On the other hand, [AV]root-security needs very little work memory on its own, so even if it was already somewhat tight in that area, which it isn't, it wouldn't matter to make it tighter. Correct me if I'm wrong, but it seems to me that [AV]sitA/B need no changes at all, so memory is not a concern for them.

Another possibility, which I am actually leaning toward, is to simply revert the change that implements a security check for 90005 in [AV]menu, and add a message when the security setting is changed, to let the users implement their own security check by monitoring it. One valid answer to "why is it ignoring my security setting" is "because scripts are supposed to check for security themselves". This message is simple to implement, simple for third parties to monitor and honour, and can be implemented in [AV]root-security just as well.
